README
NAME
    Alien::IUP - Building, finding and using iup + related libraries -
    <http://www.tecgraf.puc-rio.br/iup/>

VERSION
    Version 0.708 of Alien::IUP is based on the following:

    *   *iup* library 3.19.1 - see <http://www.tecgraf.puc-rio.br/iup/>

    *   *im* library 3.11 - see <http://www.tecgraf.puc-rio.br/im/>

    *   *cd* library 5.10 - see <http://www.tecgraf.puc-rio.br/cd/>

SYNOPSIS
    IMPORTANT: This module is not a perl binding for *iup + related*
    libraries; it is just a helper module. The real perl binding is
    implemented by IUP module, which is using Alien::IUP to locate *iup +
    related* libraries on your system (or build it from source codes).

    Alien::IUP installation comprise of:

    *   Downloading *iup* & co. source code tarballs

    *   Building *iup* & co. binaries from source codes (note: static
        libraries are build)

    *   Installing libs and dev files (*.h, *.a) into *share* directory of
        Alien::IUP distribution - *share* directory is usually something
        like this: /usr/lib/perl5/site_perl/5.18/auto/share/dist/Alien-IUP

    Later on you can use Alien::IUP in your module that needs to link with
    *iup* and/or related libs like this:

     # Sample Makefile.pl
     use ExtUtils::MakeMaker;
     use Alien::IUP;
 
     WriteMakefile(
       NAME         => 'Any::IUP::Module',
       VERSION_FROM => 'lib/Any/IUP/Module.pm',
       LIBS         => Alien::IUP->config('LIBS'),
       INC          => Alien::IUP->config('INC'),
       # + additional params
     );

    IMPORTANT: As Alien::IUP builds static libraries the modules using
    Alien::IUP (e.g. IUP) need to have Alien::IUP just for building, not for
    later use. In other words Alien:IUP is just "build dependency" not
    "run-time dependency".

METHODS
  config()
    This function is the main public interface to this module.

     Alien::IUP->config('LIBS');

    Returns a string like: '-L/path/to/iupdir/lib -liup -lim -lcd'

     Alien::IUP->config('INC');

    Returns a string like: '-I/path/to/iupdir/include'

     Alien::IUP->config('PREFIX');

    Returns a string like: '/path/to/iupdir' (note: if using the already
    installed *iup* config('PREFIX') returns undef)

  havelib()
    Checks the presence of given iup related libraries.

     Alien::IUP->havelib('iupim');
     #or
     Alien::IUP->havelib('iupim', 'iupcd', 'iupcontrols');

    Parameter(s): One or more iup related lib names - e.g. iup, cd, im,
    iupcd, iupim, iupcontrols, iup_pplot, iupimglib, iupgl, iupole.

    Returns: 1 if all libs specified as a param are available; 0 otherwise.

AUTHOR
    KMX, <kmx at cpan.org>

BUGS
    Please report any bugs or feature requests to "bug-alien-iup at
    rt.cpan.org", or through the web interface at
    <http://rt.cpan.org/NoAuth/Bugs.html?Dist=Alien-IUP>.

LICENSE AND COPYRIGHT
    Libraries *iup*, *im* and *cd*: Copyright (C) 1994-2015 Tecgraf,
    PUC-Rio. <http://www.tecgraf.puc-rio.br>

    Alien::IUP module: Copyright (C) 2015 KMX.

    This program is distributed under the MIT License:
    <http://www.opensource.org/licenses/mit-license.php>

    Permission is hereby granted, free of charge, to any person obtaining a
    copy of this software and associated documentation files (the
    "Software"), to deal in the Software without restriction, including
    without limitation the rights to use, copy, modify, merge, publish,
    distribute, sublicense, and/or sell copies of the Software, and to
    permit persons to whom the Software is furnished to do so, subject to
    the following conditions:

    The above copyright notice and this permission notice shall be included
    in all copies or substantial portions of the Software.

    THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
    OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
    MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
    IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
    CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
    TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
    SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.



Changes
Revision history for perl module Alien-IUP

0.709 2016-09-07
    -  Win32/mingw build fix

0.708 2016-09-05
    -  UNIX build fix (when GL libs/headers nto found)
    -  Cygwin fixes

0.707 2016-09-04
    -  UNIX build fix (ftgl related)

0.706 2016-07-13
    -  MS Win build fix

0.705 2016-07-10
    -  IMPORTANT: switch to iup-3.19.1+im-3.11+cd-5.10

0.704 2015-06-04
    -  IMPORTANT: switch to iup-3.14+im-3.9.1+cd-5.8.2

0.703 2015-02-08
    -  improved checking of OpenGL libraries (UNIX)

0.702 2015-02-06
    -  cygwin related patches

0.701 2015-02-06
    -  IMPORTANT: switch to iup-3.13+im-3.9+cd-5.8.1
    -  new: iupglcontrols iup_scintilla cdcontextplus
    -  fix for: https://rt.cpan.org/Public/Bug/Display.html?id=94615

0.601 2013-12-10
    -  IMPORTANT: switch to iup-3.9+im-3.8.2+cd-5.7
    -  tempfile related fix

0.501 2013-02-04
    -  skipping iup_mglplot on MSWin32+gcc3
       http://www.cpantesters.org/cpan/report/51d1df34-7637-1014-9907-a394d871b27e
    -  patch/hack: iup_mglplot + solaris 5.11/11.0 + gcc-4.5.2 
       http://www.cpantesters.org/cpan/report/187c47b6-6db2-11e2-8782-3cee870e5ae2
    -  IMPORTANT: switch to iup-3.7+im-3.8.1+cd-5.6

0.401 2012-12-31
    -  IMPORTANT: switch to iup-3.6 (new iup_mglplot introduced)

0.305 2011-08-20
    -  iup-imglib patch
    -  cd patch (crutial - without this patch IUP module segfaults at some points)

0.304 2011-08-19
    -  cygwin patch (__argc + __argv related)

0.303 2011-08-08
    -  added support for -DIUP_IMGLIB_LARGE (now on Win32 only)
    -  fixed http://www.cpantesters.org/cpan/report/889253e2-b5f8-11e0-8204-0f92f06c0dc8 (kFreeBSD failure)
    -  better diagnostic messages

0.302 2011-07-22
    -  solaris workaround (ar, ranlib)

0.301 2011-05-12
    -  IMPORTANT: switch to iup-3.5

0.204 2011-04-05
    -  IMPORTANT: the last release with iup-3.4
    -  back from Digest::file to Digest::SHA

0.203 2011-04-04
    -  fixed "missing download dir" issue

0.202 2011-04-03
    -  fixed http://www.cpantesters.org/cpan/report/61d37376-471c-11e0-b0ff-ff1c7a0d71a6 (File::Spec & devnull)
    -  using LWP::Simple instead of File::Fetch for download
    -  using Digest::file instead of Digest::SHA

0.201 2011-02-18
    -  IMPORTANT: the first stable release with iup-3.4
    -  dev build: switch to iup-3.4 + updated im, cd

0.116 2011-02-10
    -  IMPORTANT: the last release with iup-3.3
    -  Unix: if not enough libs found - show more debug info

0.115 2011-02-03
    -  Alien::IUP->havelib() introduced
    -  preinstalled lib detection disabled (untill somebody will ask for it)

0.114 2011-01-27
    -  not building libs that are currently not used by IUP
    -  less build messages by default
    -  dev build: with the latest CVS sources

0.113 2011-01-05
    -  same as 0.112_01 + updated TODO
    -  fixing -lstdc++ issue on non-gcc compilers

0.112 2010-12-30
    -  skipping iuptuio on solaris - http://www.cpantesters.org/cpan/report/c39d9264-0ae3-11e0-afb0-adca6bb533f3
    -  skipping iuptuio on cygwin (not only gcc3)

0.111 2010-12-16
    -  GTK/GTK_BASE workaround fixing http://www.cpantesters.org/cpan/report/d67a6578-08aa-11e0-afb0-adca6bb533f3

0.110 2010-12-13
    -  iup-gtk-tabs.diff fixing http://www.cpantesters.org/cpan/report/b524182a-04c1-11e0-afb0-adca6bb533f3
    -  fixing gtkprintunixdialog.h failure http://www.cpantesters.org/cpan/report/9629bbba-0655-11e0-9ee0-f20c722eace1

0.109 2010-12-12
    -  fixing im-INC-order.diff + cd-GTK_BASE.diff

0.108 2010-12-09_01+0.108_02
    -  dev build: with the latest CVS sources

0.108 2010-12-08
    -  info.t fix

0.107 2010-12-07
    -  iuptuio target added (Unix.pm)
    -  more debug info saved to config_data
    -  updated im-INC-order.diff

0.106 2010-12-02
    -  updated cd-GTK_BASE.diff

0.105 2010-12-02
    -  important fix: GTK2 + pkg-config
    -  improved build target autoselection

0.104 2010-12-01
    -  updated iup-GTK_WIDGET.diff

0.103 2010-12-01
    -  Motif/X11 build fix
    -  fix for http://www.cpantesters.org/cpan/report/ec6300d2-fb91-11df-a694-e2b03d723d5a

0.102 2010-12-01
    -  patches: cd-GTK_BASE.diff, iup-GTK_WIDGET.diff
    -  linking lib order fixed
    -  showing mod version in debug info

0.101 2010-11-23
    -  IMPORTANT: switch to IUP 3.3 (+ upgrade of other libs as well)
    -  NetBSD ntohl() related patch (#include <sys/types.h>)
    -  significant Windows build changes
    -  warn (not skip) by im_process on OpenBSD
    -  do not build freetype if we halready have freetype on our system
    -  ms64bit build patch
    -  more diagnostics (freetype)
    -  require installed cairo for GTK2 build
    -  fixing netbsd failure (using USE_PKGCONFIG)
    -  webkit/webkit.h test (disabling iupweb if no webkit)
    -  patches for DragonFly BSD
    -  improved cygwin build
    -  new checksum for updated iup-3.3_Sources.tar.gz
    -  new builder for cygwin (separated from generic UNIX)

0.024 2010-11-14
    -  IMPORTANT: the last release with iup-3.2
    -  die if no GUI lib detected
    -  switching Changes fromat from plain text to POD
    -  changing version numbering

0.0.23 2010-10-20
    -  create_build_script moved after GNU make check

0.0.22 2010-10-12
    -  fixed GNU make check exit code

0.0.21 2010-10-11
    -  GNU make check moved to Build.pl (now cpan testers should report "N/A)
    -  more messages on sha1 check failure
    -  small POD changes

0.0.20 2010-10-05
    -  more questions/prompts during build
    -  cleaning up + cosmetics
    -  OpenBSD - not building im_process (g++ internal error)
    -  OpenBSD - lesstif related fixes
    -  DragonFly patch for 'sizeof(FILE)'
    -  exit if GNU make not found
    -  DragonFly patch for libjasper/jasper/jas_types.h
    -  create_makefile_pl => 'small' (creating Makefile.PL to satisfy old CPAN clients)

lib/Alien/IUP.pm
package Alien::IUP;

use warnings;
use strict;

use Alien::IUP::ConfigData;
use File::ShareDir qw(dist_dir);
use File::Spec::Functions qw(catdir catfile rel2abs);

=head1 NAME

Alien::IUP - Building, finding and using iup + related libraries - L<http://www.tecgraf.puc-rio.br/iup/>

=cut

our $VERSION = "0.709";

=head1 VERSION

=over

=item * I<iup> library 3.19.1 - see L<http://www.tecgraf.puc-rio.br/iup/>

=item * I<im> library 3.11 - see L<http://www.tecgraf.puc-rio.br/im/>

=item * I<cd> library 5.10 - see L<http://www.tecgraf.puc-rio.br/cd/>

=back

=head1 SYNOPSIS

B<IMPORTANT:> This module is not a perl binding for I<iup + related> libraries; it is just
a helper module. The real perl binding is implemented by L<IUP|IUP> module,
which is using Alien::IUP to locate I<iup + related> libraries on your system (or build it from source codes).

Alien::IUP installation comprise of:

=over

=item * Downloading I<iup> & co. source code tarballs

=item * Building I<iup> & co. binaries from source codes (note: static libraries are build)

=item * Installing libs and dev files (*.h, *.a) into I<share> directory of Alien::IUP
distribution - I<share> directory is usually something like this: /usr/lib/perl5/site_perl/5.18/auto/share/dist/Alien-IUP

=back

Later on you can use Alien::IUP in your module that needs to link with
I<iup> and/or related libs like this:

 # Sample Makefile.pl
 use ExtUtils::MakeMaker;
 use Alien::IUP;
 
 WriteMakefile(
   NAME         => 'Any::IUP::Module',
   VERSION_FROM => 'lib/Any/IUP/Module.pm',
   LIBS         => Alien::IUP->config('LIBS'),
   INC          => Alien::IUP->config('INC'),
   # + additional params
 );

B<IMPORTANT:> As Alien::IUP builds static libraries the modules using Alien::IUP (e.g. L<IUP|IUP>)
need to have Alien::IUP just for building, not for later use. In other words Alien:IUP is just
"build dependency" not "run-time dependency".

=head1 METHODS

=head2 config()

This function is the main public interface to this module.

 Alien::IUP->config('LIBS');

Returns a string like: '-L/path/to/iupdir/lib -liup -lim -lcd'

 Alien::IUP->config('INC');

Returns a string like: '-I/path/to/iupdir/include'

 Alien::IUP->config('PREFIX');

Returns a string like: '/path/to/iupdir' (note: if using the already installed
I<iup> config('PREFIX') returns undef)

=head2 havelib()

Checks the presence of given iup related libraries.

 Alien::IUP->havelib('iupim');
 #or
 Alien::IUP->havelib('iupim', 'iupcd', 'iupcontrols');

Parameter(s): One or more iup related lib names - e.g. iup, cd, im, iupcd, iupim, iupcontrols, iup_pplot, iupimglib, iupgl, iupole.

Returns: 1 if all libs specified as a param are available; 0 otherwise.

=head1 AUTHOR

KMX, E<lt>kmx at cpan.orgE<gt>

=head1 BUGS

Please report any bugs or feature requests to C<bug-alien-iup at rt.cpan.org>, or through
the web interface at L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=Alien-IUP>.

=head1 LICENSE AND COPYRIGHT

Libraries I<iup>, I<im> and I<cd>: Copyright (C) 1994-2015 Tecgraf, PUC-Rio.
L<http://www.tecgraf.puc-rio.br>

Alien::IUP module: Copyright (C) 2015 KMX.

This program is distributed under the MIT License:
L<http://www.opensource.org/licenses/mit-license.php>

Permission is hereby granted, free of charge, to any person
obtaining a copy of this software and associated documentation
files (the "Software"), to deal in the Software without
restriction, including without limitation the rights to use,
copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the
Software is furnished to do so, subject to the following
conditions:

The above copyright notice and this permission notice shall be
included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES
OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,
WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
OTHER DEALINGS IN THE SOFTWARE.

=cut

sub config {
  my ($package, $param) = @_;
  return unless ($param =~ /[a-z0-9_]*/i);
  my $subdir = Alien::IUP::ConfigData->config('share_subdir');
  unless ($subdir) {
    #we are using lib already installed on your system not compiled by Alien
    #therefore no additional magic needed
    return Alien::IUP::ConfigData->config('config')->{$param};
  }
  my $share_dir = dist_dir('Alien-IUP');
  my $real_prefix = catdir($share_dir, $subdir);
  my $val = Alien::IUP::ConfigData->config('config')->{$param};
  return unless $val;
  $val =~ s/\@PrEfIx\@/$real_prefix/g; # handle @PrEfIx@ replacement
  return $val;
}

sub havelib {
  my ($package, @libs) = @_;
  my $iuplibs = my $subdir = Alien::IUP::ConfigData->config('iup_libs');
  return 0 unless defined $iuplibs;
  for (@libs) {
    return 0 unless defined $iuplibs->{$_} && $iuplibs->{$_} == 1;
  }
  return 1;
}

1; # End of Alien::IUP
